drivers/serial/uart_npcx.c

```diff
@@ -0,0 +1,376 @@
+ */
+
+#define DT_DRV_COMPAT nuvoton_npcx_uart
+
+#include <soc.h>
+#include <drivers/uart.h>
+#include <drivers/clock_control.h>
+#include <assert.h>
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(uart_npcx, LOG_LEVEL_ERR);
+
+/* Device config */
+struct uart_npcx_config {
+	struct uart_device_config uconf;
+	/* clock configuration */
+	struct npcx_clk_cfg clk_cfg;
+	/* pinmux configuration */
+	const uint8_t   alts_size;
+	const struct npcx_alt *alts_list;
+};
+
+/* Driver data */
+struct uart_npcx_data {
+	/* Baud rate */
+	uint32_t baud_rate;
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+	uart_irq_callback_user_data_t user_cb;
+	void *user_data;
+#endif
+};
+
+/* Driver convenience defines */
+#define DRV_CONFIG(dev) \
+	((const struct uart_npcx_config *)(dev)->config_info)
+
+#define DRV_DATA(dev) \
+	((struct uart_npcx_data *)(dev)->driver_data)
+
+#define HAL_INSTANCE(dev) \
+	(uart_inst_t)(DRV_CONFIG(dev)->uconf.base)
+
+/******************************************************************************/
+/* UART local functions */
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static void uart_npcx_dis_all_tx_interrupts(struct device *dev)
+{
+	uart_inst_t inst = HAL_INSTANCE(dev);
+
+	/* Disable all Tx interrupts */
+	NPCX_UFTCTL(inst) &= ~(BIT(NPCX_UFTCTL_TEMPTY_LVL_EN) |
+				BIT(NPCX_UFTCTL_TEMPTY_EN) |
+				BIT(NPCX_UFTCTL_NXMIPEN));
+}
+
+static void uart_npcx_clear_rx_fifo(struct device *dev)
+{
+	uart_inst_t inst = HAL_INSTANCE(dev);
+	uint8_t scratch;
+
+	/* Read all dummy bytes out in Rx FIFO */
+	while (NPCX_UART_RX_FIFO_IS_AVAILABLE(inst))
+		scratch = NPCX_URBUF(inst);
+}
+#endif
+/******************************************************************************/
+/* UART api functions */
+static int uart_npcx_poll_in(struct device *dev, unsigned char *c)
+{
+	uart_inst_t inst = HAL_INSTANCE(dev);
+
+	/* Rx reg is not full. (Single byte mode) */
+	if (!IS_BIT_SET(NPCX_UICTRL(inst), NPCX_UICTRL_RBF))
+		return -1;
+
+	*c = NPCX_URBUF(inst);
+	return 0;
+}
+
+static void uart_npcx_poll_out(struct device *dev,
+					 unsigned char c)
+{
+	uart_inst_t inst = HAL_INSTANCE(dev);
+
+	/* Wait while tx reg is ready to send. (Single byte mode) */
+	while (!IS_BIT_SET(NPCX_UICTRL(inst), NPCX_UICTRL_TBE))
+		;
+
+	NPCX_UTBUF(inst) = c;
+}
+
+static int uart_npcx_err_check(struct device *dev)
+{
+	uart_inst_t inst = HAL_INSTANCE(dev);
+	uint32_t err = 0U;
+	uint8_t stat = NPCX_USTAT(inst);
+
+	if (IS_BIT_SET(stat, NPCX_USTAT_DOE))
+		err |= UART_ERROR_OVERRUN;
+
+	if (IS_BIT_SET(stat, NPCX_USTAT_PE))
+		err |= UART_ERROR_PARITY;
+
+	if (IS_BIT_SET(stat, NPCX_USTAT_FE))
+		err |= UART_ERROR_FRAMING;
+
+	return err;
+}
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static int uart_npcx_fifo_fill(struct device *dev, const uint8_t *tx_data,
+				  int size)
+{
+	uart_inst_t inst = HAL_INSTANCE(dev);
+	uint8_t tx_bytes = 0U;
+
+	/* If Tx FIFO is still ready to send */
+	while ((size - tx_bytes > 0) && NPCX_UART_TX_FIFO_IS_READY(inst)) {
```

Comment:
I presume that NPCX_UART_TX_FIFO_IS_READY is defined by the HAL

I really wonder whether it would be better to have the code all here and drop the HAL

Reply:
yes please to dropping the hal